File: src/java/org/apache/fop/datatypes/LengthBase.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* $Id$ */

package org.apache.fop.datatypes;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import org.apache.fop.fo.Constants;
import org.apache.fop.fo.FObj;
import org.apache.fop.fo.PropertyList;
import org.apache.fop.fo.expr.PropertyException;
import org.apache.fop.util.CompareUtil;

/**
 * Models a length which can be used as a factor in a percentage length
 * calculation
 */
public class LengthBase implements PercentBase {
    // Standard kinds of percent-based length
    /** constant for a custom percent-based length */
    public static final int CUSTOM_BASE = 0;
    /** constant for a font-size percent-based length */
    public static final int FONTSIZE = 1;
    /** constant for an inh font-size percent-based length */
    public static final int INH_FONTSIZE = 2;
    /** constant for a containing box percent-based length */
    public static final int PARENT_AREA_WIDTH = 3;
    /** constant for a containing refarea percent-based length */
    public static final int CONTAINING_REFAREA_WIDTH = 4;
    /** constant for a containing block percent-based length */
    public static final int CONTAINING_BLOCK_WIDTH = 5;
    /** constant for a containing block percent-based length */
    public static final int CONTAINING_BLOCK_HEIGHT = 6;
    /** constant for a image intrinsic percent-based length */
    public static final int IMAGE_INTRINSIC_WIDTH = 7;
    /** constant for a image intrinsic percent-based length */
    public static final int IMAGE_INTRINSIC_HEIGHT = 8;
    /** constant for a image background position horizontal percent-based length */
    public static final int IMAGE_BACKGROUND_POSITION_HORIZONTAL = 9;
    /** constant for a image background position vertical percent-based length */
    public static final int IMAGE_BACKGROUND_POSITION_VERTICAL = 10;
    /** constant for a table-unit-based length */
    public static final int TABLE_UNITS = 11;
    /** constant for a alignment adjust percent-based length */
    public static final int ALIGNMENT_ADJUST = 12;

    /**
     * logging instance
     */
    protected static final Log log = LogFactory.getLog(LengthBase.class);

    /**
     * The FO for which this property is to be calculated.
     */
    protected /* final */ FObj fobj;

    /**
     * One of the defined types of LengthBase
     */
    private /* final */ int baseType;

    /** For percentages based on other length properties */
    private Length baseLength;

    /**
     * Constructor
     * @param plist property list for this
     * @param baseType a constant defining the type of teh percent base
     * @throws PropertyException In case an problem occurs while evaluating values
     */
    public LengthBase(PropertyList plist, int baseType) throws PropertyException {
        this.fobj = plist.getFObj();
        this.baseType = baseType;
        switch (baseType) {
        case FONTSIZE:
            this.baseLength = plist.get(Constants.PR_FONT_SIZE).getLength();
            break;
        case INH_FONTSIZE:
            this.baseLength = plist.getInherited(Constants.PR_FONT_SIZE).getLength();
            break;
        default:
            // TODO: pacify CheckStyle
            // throw new RuntimeException();
            break;
        }
    }

    /**
     * @return the dimension of this object (always 1)
     */
    public int getDimension() {
        return 1;
    }

    /**
     * @return the base value of this object (always 1.0)
     */
    public double getBaseValue() {
        return 1.0;
    }

    /**
     * {@inheritDoc}
     */
    public int getBaseLength(PercentBaseContext context) throws PropertyException {
        int baseLen = 0;
        if (context != null) {
            if (baseType == FONTSIZE || baseType == INH_FONTSIZE) {
                return baseLength.getValue(context);
            }
            baseLen =  context.getBaseLength(baseType,  fobj);
        } else {
            log.error("getBaseLength called without context");
        }
        return baseLen;
    }

    /** {@inheritDoc} */
    public String toString() {
        return super.toString()
          + "[fo=" + fobj + ","
          + "baseType=" + baseType + ","
          + "baseLength=" + baseLength + "]";
    }

    /**@return the base length as a {@link Length} */
    public Length getBaseLength() {
        return baseLength;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + CompareUtil.getHashCode(baseLength);
        result = prime * result + baseType;
        result = prime * result + CompareUtil.getHashCode(fobj);
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof LengthBase)) {
            return false;
        }
        LengthBase other = (LengthBase) obj;
        return CompareUtil.equal(baseLength, other.baseLength)
                && baseType == other.baseType
                && CompareUtil.equal(fobj, other.fobj);
    }

}

File: dotnet/src/Fop.DataTypes/LengthBase.cs
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using Fop.Fo.Expr;

namespace Fop.DataTypes;

/// <summary>
/// Models a length which can be used as a factor in a percentage length calculation.
/// <para>
/// Port of <c>org.apache.fop.datatypes.LengthBase</c>. The original primary constructor takes a
/// <c>PropertyList</c> and reads the (inherited) font-size property from it; since the Fop.Fo
/// property layer is not ported yet, this port instead takes the FO object, the base type, and an
/// optional resolved base length directly.
/// </para>
/// <para>TODO: replace the object fobj parameter and add the PropertyList-based construction when Fop.Fo is ported.</para>
/// </summary>
public class LengthBase : IPercentBase
{
    // Standard kinds of percent-based length.

    /// <summary>Constant for a custom percent-based length.</summary>
    public const int CustomBase = 0;

    /// <summary>Constant for a font-size percent-based length.</summary>
    public const int FontSize = 1;

    /// <summary>Constant for an inherited font-size percent-based length.</summary>
    public const int InhFontSize = 2;

    /// <summary>Constant for a containing-box percent-based length.</summary>
    public const int ParentAreaWidth = 3;

    /// <summary>Constant for a containing reference-area percent-based length.</summary>
    public const int ContainingRefAreaWidth = 4;

    /// <summary>Constant for a containing-block percent-based length (width).</summary>
    public const int ContainingBlockWidth = 5;

    /// <summary>Constant for a containing-block percent-based length (height).</summary>
    public const int ContainingBlockHeight = 6;

    /// <summary>Constant for an image-intrinsic percent-based length (width).</summary>
    public const int ImageIntrinsicWidth = 7;

    /// <summary>Constant for an image-intrinsic percent-based length (height).</summary>
    public const int ImageIntrinsicHeight = 8;

    /// <summary>Constant for an image background-position horizontal percent-based length.</summary>
    public const int ImageBackgroundPositionHorizontal = 9;

    /// <summary>Constant for an image background-position vertical percent-based length.</summary>
    public const int ImageBackgroundPositionVertical = 10;

    /// <summary>Constant for a table-unit-based length.</summary>
    public const int TableUnits = 11;

    /// <summary>Constant for an alignment-adjust percent-based length.</summary>
    public const int AlignmentAdjust = 12;

    /// <summary>The FO for which this property is to be calculated.</summary>
    protected object? fobj;

    private readonly int baseType;

    /// <summary>For percentages based on other length properties.</summary>
    private readonly ILength? baseLength;

    /// <summary>Creates a new length base.</summary>
    /// <param name="fobj">The FO object for which the percentage is to be calculated.</param>
    /// <param name="baseType">One of the defined <c>LengthBase</c> type constants.</param>
    /// <param name="baseLength">
    /// The resolved base length, used for the <see cref="FontSize"/>/<see cref="InhFontSize"/> base types.
    /// </param>
    public LengthBase(object? fobj, int baseType, ILength? baseLength = null)
    {
        this.fobj = fobj;
        this.baseType = baseType;
        this.baseLength = baseLength;
    }

    /// <summary>The dimension of this object (always 1).</summary>
    public int Dimension => 1;

    /// <summary>The base value of this object (always 1.0).</summary>
    public double BaseValue => 1.0;

    /// <summary>The base length as an <see cref="ILength"/>.</summary>
    public ILength? BaseLength => baseLength;

    /// <summary>Returns the base length value in millipoints for the given context.</summary>
    /// <exception cref="PropertyException">If a problem occurs during evaluation of this value.</exception>
    public int GetBaseLength(IPercentBaseContext? context)
    {
        int baseLen = 0;
        if (context != null)
        {
            if (baseType is FontSize or InhFontSize)
            {
                // baseLength is expected to be non-null for the font-size base types.
                return baseLength!.GetValue(context);
            }

            baseLen = context.GetBaseLength(baseType, fobj);
        }
        else
        {
            // The Java original logs an error ("getBaseLength called without context") here;
            // logging is omitted in this standalone port but the return value (0) is preserved.
        }

        return baseLen;
    }

    /// <summary>Returns a string representation of this length base.</summary>
    public override string ToString() =>
        $"{base.ToString()}[fo={fobj},baseType={baseType},baseLength={baseLength}]";

    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(baseLength, baseType, fobj);

    /// <inheritdoc/>
    public override bool Equals(object? obj)
    {
        if (ReferenceEquals(this, obj))
        {
            return true;
        }

        return obj is LengthBase other
            && Equals(baseLength, other.baseLength)
            && baseType == other.baseType
            && Equals(fobj, other.fobj);
    }
}
